FirebaseAppDistribution/Sources/FIRAppDistributionAuthPersistence.m

```diff
@@ -16,77 +16,99 @@
+    *error = [NSError errorWithDomain:kFIRAppDistributionKeychainErrorDomain
+                                 code:FIRAppDistributionErrorTokenDeletionFailure
+                             userInfo:userInfo];
+  }
+  return errSecSuccess || status == errSecItemNotFound;
 }
 
-+ (OIDAuthState*)retrieveAuthState {
-    NSMutableDictionary *keychainQuery = [self getKeyChainQuery];
-      [keychainQuery setObject:(id)kCFBooleanTrue forKey:(id)kSecReturnData];
-      [keychainQuery setObject:(id)kSecMatchLimitOne forKey:(id)kSecMatchLimit];
-      CFDataRef passwordData = NULL;
-      NSData *result = nil;
-      OSStatus status = SecItemCopyMatching((CFDictionaryRef)keychainQuery,
-                                         (CFTypeRef *)&passwordData);
-      if (status == noErr && 0 < [(__bridge NSData *)passwordData length]) {
-        result = [(__bridge NSData *)passwordData copy];
-      } else {
-          NSLog(@"AUTH ERROR - cannot lookup keystore config");
-      }
-      if (passwordData != NULL) {
-        CFRelease(passwordData);
-      }
-    
-    OIDAuthState *authState = nil;
-    if(result) {
-        authState = (OIDAuthState *)[NSKeyedUnarchiver unarchiveObjectWithData:result];
-    }
-    
-    return authState;
++ (OIDAuthState *)retrieveAuthState:(NSError **_Nullable)error {
+  NSMutableDictionary *keychainQuery = [self getKeyChainQuery];
+  [keychainQuery setObject:(id)kCFBooleanTrue forKey:(id)kSecReturnData];
+  [keychainQuery setObject:(id)kSecMatchLimitOne forKey:(id)kSecMatchLimit];
+  CFDataRef passwordData = NULL;
+  NSData *result = nil;
+  OSStatus status = SecItemCopyMatching((CFDictionaryRef)keychainQuery, (CFTypeRef *)&passwordData);
+
+  if (status == noErr && 0 < [(__bridge NSData *)passwordData length]) {
+    result = [(__bridge NSData *)passwordData copy];
+  } else if (error) {
+    NSString *desc = NSLocalizedString(
+        @"Failed to retrieve auth state from keychain. Tester will have to sign in again.",
+        @"Error message for failure to retrieve auth state from keychain");
+    NSDictionary *userInfo = @{NSLocalizedDescriptionKey : desc};
+    *error = [NSError errorWithDomain:kFIRAppDistributionKeychainErrorDomain
+                                 code:FIRAppDistributionErrorTokenRetrievalFailure
+                             userInfo:userInfo];
+  }
+
+  OIDAuthState *authState = nil;
+  if (result) {
+    authState = (OIDAuthState *)[NSKeyedUnarchiver unarchiveObjectWithData:result];
+  } else if (error) {
+    NSString *desc =
+        NSLocalizedString(@"Failed to unarchive auth state. Tester will have to sign in again.",
+                          @"Error message for failure to copy password data from keychain");
+    NSDictionary *userInfo = @{NSLocalizedDescriptionKey : desc};
+    *error = [NSError errorWithDomain:kFIRAppDistributionKeychainErrorDomain
+                                 code:FIRAppDistributionErrorTokenRetrievalFailure
+                             userInfo:userInfo];
+  }
```

Comment:
To me it's odd that this will overwrite the error you built on line 54 every time result is null